### PR TITLE
Add 'meson subprojects update' command

### PR DESCRIPTION
**Pull request by xclaesse**
This is inspired by gst-build's git-update script.

### CODE REVIEW

File: mesonbuild/msubprojects.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+from . import mlog
+from .wrap.wrap import PackageDefinition
+
+def git(cmd, workingdir):
+    return subprocess.check_output(['git', '-C', workingdir] + cmd,
+                                   stderr=subprocess.STDOUT).decode()
+
+def update_git(wrap, repo_dir, options):
+    revision = wrap.get('revision')
+    git(['config', 'rebase.autoStash', 'true'], repo_dir)
```
Comment (jpakkane) [inline mesonbuild/msubprojects.py]: Changing the user's Git settings like this is not allowed under any circumstances!

Comment (xclaesse) [reply mesonbuild/msubprojects.py]: I actually agree. I'll change that by doing a 'git stash' and 'git stash pop' explicitly, which has the extra advantage of also working in the case `revision` is a tag that we checkout. `rebase.autoStash` trick only works in the case `revision` is a branch.

**Comment by jpakkane** [comment]
The wrap command already has an `update` command. Maybe we could have `updateall` there instead of adding a new top level command?

**Comment by xclaesse** [comment]
You mean wraptool.update() ? It seems to be unrelated.

**Comment by jpakkane** [comment]
I meant `meson wrap update subprojname`. Granted it currently (IIRC) looks up stuff from wrapdb, but it could detect that it is being run on a Git repo and do a pull.

**Comment by xclaesse** [comment]
The CI error is unrelated:
```
meson.build:1:0: ERROR:  Value "c++17" for combo option "cpp_std" is not one of the choices. Possible choices are: "none", "c++11", "c++14", "c++latest".
```

@jpakkane does it look good like that?

**Comment by xclaesse** [comment]
@thiblahute since you did the git-update script, would be nice to have your opinion here ;-)

**Comment by xclaesse** [comment]
@jpakkane any objection on this?

File: mesonbuild/msubprojects.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
+import os, subprocess, argparse
```
Comment (ignatenkobrain) [inline mesonbuild/msubprojects.py]: __[Flake8]__

[F401] 'argparse' imported but unused

**Comment by xclaesse** [comment]
@jpakkane FYI the use-case for `meson subprojects download` is for gst-build CI. gst-build is a meta project, it has every GStreamer components as subproject. For the CI we want to fetch gst-build and its subprojects while generating the docker image, like that for each build job we only have to update the git repository instead of cloning them all from scratch. Currently we do that with basically `meson builddir && rm -rf builddir` but that run a useless configure just to fetch subprojects, and also we need different fallback subprojects depending on which type of build we do (native build, cross for android, etc) so some subprojects are missing.

I think it's generally useful to be able to split the download and configure steps. If you do `meson subprojects download` then you can later build offline with `meson --wrap-mode=nodownload builddir`. Thinking about this, we could even have a "dist" mode that then include them all into a giant tarball to have perfectly reproducible offline builds, but one step at a time.

**Comment by jpakkane** [comment]
As a feature this is useful and definitely something we want to provide. It's just a question of polishing the UI. It's a bit unfortunate that this means there will be two different ways of updating things, one for wrapdb and a second one for other subproject types. I can't really come up with a better way to split these things, though.

This requires the requisite documentation updates, including something in the actual manual showing usage examples. It would also be nice to get some tests, but since this is mostly about network operations it might be tricky.

**Comment by xclaesse** [comment]
@jpakkane I don't think the CI failure is related, and I don't understand how to get from azure where is the error.

File: mesonbuild/msubprojects.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+    else:
+        mlog.log('  -> Cannot update', wrap.type, 'subproject')
+
+def checkout(wrap, repo_dir, options):
+    if wrap.type != 'git' or not os.path.isdir(repo_dir):
+        return
+    branch_name = options.branch_name if options.branch_name else wrap.get('revision')
+    mlog.log('Checkout %s in %s...' % (branch_name, wrap.name))
+    try:
+        git(['checkout', '-B', branch_name, '--'], repo_dir)
```
Comment (textshell) [inline mesonbuild/msubprojects.py]: I think (from reading the docs) `-B` will clobber an existing branch. that does not sound like "checkout" to me.

File: mesonbuild/msubprojects.py [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+                                               'This can be used to pre-fetch all subprojects and avoid downloads during configure.')
+    add_common_arguments(p)
+    p.set_defaults(subprojects_func=download)
+
+def run(options):
+    src_dir = os.path.relpath(os.path.realpath(options.sourcedir))
+    if not os.path.isfile(os.path.join(src_dir, 'meson.build')):
+        mlog.error('Directory', mlog.bold(src_dir), 'does not seem to be a Meson source directory.')
+        return 1
+    subprojects_dir = os.path.join(src_dir, 'subprojects')
```
Comment (textshell) [inline mesonbuild/msubprojects.py]: I think the path to subprojects is still configurable?

File: setup.cfg [hunk 5]
```diff
@@ -26,4 +26,6 @@ ignore =
     E741
     # E722: do not use bare except'
     E722
+    # W504: line break after binary operator
+    W504
```
Comment (textshell) [inline setup.cfg]: do we really want this? Seems like a code style thing we would not want to be just an after-though in some PR.

File: mesonbuild/msubprojects.py (hunk 3 above)
Comment (xclaesse) [reply mesonbuild/msubprojects.py]: Oops you're right, that isn't the intention. I could just check if the branch exists before, but I wanted to profit from `git checkout foo` feature that create a branch if it exists in a remote. But I can't find a solution for that, so I think I'll have to add `meson subprojects checkout -b <branch>`... I'll think about that a bit more.

File: setup.cfg (hunk 5 above)
Comment (xclaesse) [reply setup.cfg]: My flake8 doesn't complain about it, but without that the CI complains for code I didn't touch in this PR :(

File: mesonbuild/msubprojects.py (hunk 4 above)
Comment (xclaesse) [reply mesonbuild/msubprojects.py]: Yes, but AFAIK it's deprecated for a long time, and I would like this command-line tool to work without a build dir (before configuration) so I cannot know if that option is set to a different value.

Comment (textshell) [reply mesonbuild/msubprojects.py]: in mintro we have enough infrastructure to parse this out without a build dir. But it does not actually use it but hardcodes 'subprojects' as well. Maybe we should fix that in a later PR.

File: mesonbuild/msubprojects.py (hunk 3 above)
Comment (xclaesse) [reply mesonbuild/msubprojects.py]: Ok, updated my PR to add `-b` option to `meson subprojects checkout` that gets passed to `git checkout`.

I think the equivalent of what I wanted to do can still be done with 2 commands:
1. `meson subprojects checkout some_branch` will create branches that can be found in a remote, and fail for the subprojects that doesn't have it.
2. `meson subprojects checkout -b some_branch` will create the branch in the subprojects that failed in the first command.

I think that's good enough.

**Pull request merged**